modules/lang-painless/src/main/java/org/elasticsearch/painless/node/PSubNullSafeCallInvoke.java

```diff
@@ -52,19 +59,21 @@ void extractVariables(Set<String> variables) {
     void analyze(Locals locals) {
         guarded.analyze(locals);
         actual = guarded.actual;
-        if (actual.sort.primitive) {
-            throw new IllegalArgumentException("Result of null safe operator must be nullable");
+        if (actual.sort.primitive && defaultForNull == null) {
+            throw new IllegalArgumentException("Result of null safe dereference must be nullable unless followed by ?:");
         }
     }
 
     @Override
     void write(MethodWriter writer, Globals globals) {
         writer.writeDebugInfo(location);
 
-        Label end = new Label();
+        Label nullLabel = defaultForNull == null ? new Label() : defaultForNull.nullLabel();
         writer.dup();
-        writer.ifNull(end);
+        writer.ifNull(nullLabel);
         guarded.write(writer, globals);
-        writer.mark(end);
+        if (defaultForNull == null) {
+            writer.mark(nullLabel);
+        }
```

Comment:
It feels weird to me that this is kind of a non-local jump. I mean, it is local to the context of the `?:` operator, but it isn't normal to jump outside of the node you are working in.

On the other hand, it looks to work fairly well.